cylc/uiserver/resolvers.py

```diff
@@ -15,10 +15,190 @@
+    )
+    ret = proc.wait(timeout=5)
+    out, err = proc.communicate()
+    return ret or out.strip() == version
+
+
+class Services:
+    """Cylc services provided by the UI Server."""
+
+    @staticmethod
+    async def play(workflows, args, workflows_mgr):
+        """Calls `cylc play`."""
+        response = []
+
+        # get ready to run the command
+        try:
+            # check that the request cylc version is available
+            cylc_version = None
+            if 'cylc_version' in args:
+                cylc_version = args['cylc_version']
+                if not check_cylc_version(cylc_version):
+                    return {'error': {'message': [
+                        f'cylc version not available: {cylc_version}'
+                    ]}}
+                args = dict(args)
+                args.pop('cylc_version')
+
+            # build the command
+            cmd = ['cylc', 'play', '--color=never']
+            for key, value in args.items():
+                if value is False:
+                    # don't add binary flags
+                    continue
+                key = snake_to_kebab(key)
+                if not isinstance(value, list):
+                    value = [value]
+                for item in value:
+                    cmd.append(quote(key))
+                    if item is not True:
+                        # don't provide values for binary flags
+                        if key == '--set':
+                            # TODO
+                            # don't quote this value as it messes with things
+                            # I think subprocess quotes everything anyway
+                            # double check
+                            cmd.append(item)
+                        else:
+                            cmd.append(quote(item))
+
+        except Exception as exc:
+            # oh noes, something went wrong, send back confirmation
+            return response.append(
+                [
+                    {
+                        'error': {
+                            'message': str(exc)
+                        }
+                    }
+                ]
+            )
+
+        # start each requested flow
+        me = getpass.getuser()
+        for user, flow, _ in workflows:
+            cmd_key = f'{user}|{flow}'
+            try:
+                if user != me:
+                    # TODO: multi-user support
+                    # forward this request to the other users UIS
+                    raise ValueError(f'Cannot start flow for user: {user}')
+
+                # add the workflow to the command
+                cmd = [*cmd, quote(flow)]
+
+                # get a representation of the command being run
+                cmd_repr = ' '.join(cmd)
+                if cylc_version:
+                    cmd_repr = f'CYLC_VERSION={cylc_version} {cmd_repr}'
+                logger.info(f'$ {cmd_repr}')
+
+                # run cylc run
+                proc = Popen(
```

Comment:
This is to prevent tethering the UIS version to the Flow version allowing users to start flows with whatever Cylc versions are made available via the Cylc wrapper script.

Hopefully we will be able to push for greater uniformity in Cylc version by making it easier to upgrade UI/UIS/Flow versions on-the-fly (e.g. could add an "upgrade" service to the UIS for restarting flows under different Cylc versions), however, there is still a large use case for running with diff versions e.g. at the MO:

* Our OS/PS trailing is performed with the same Cylc version as the flows will eventually be put into production with.
* We deploy the latest Cylc release but do not make it default until we have had people try it out for a few weeks.

Hopefully the supported version range will be fairly minimal though trailing could cause a drift of say 6 months.